### PR TITLE
mute and unmute

### PR DESCRIPTION
**Pull request by qfweiwei**
## Purpose

1. Implement mute/unmute.
2. Add colorful text to highlight status and muted microphone.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: clients/cpp-console/src/windows/WindowsMicMuter.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    hr = pDevice->Activate(__uuidof(IAudioEndpointVolume), CLSCTX_INPROC_SERVER, NULL,
+        reinterpret_cast<void**>(&m_endpointVolume));
+    if (hr != S_OK)
+    {
+        goto exit;
+    }
+    m_endpointVolume->GetMute(&_originalMuteState);
+    _muted = _originalMuteState == TRUE;
```
Comment (dargilco) [inline clients/cpp-console/src/windows/WindowsMicMuter.cpp:54]: Why are the values being set to TRUE here?

File: clients/cpp-console/include/WindowsMicMuter.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#include <mmdeviceapi.h>
+#include <thread>
+#include <Windows.h>
+#include <wrl\implements.h>
+#include "MicMuter.h"
+
+//using namespace Microsoft::WRL;
+//using namespace Windows::Media::Devices;
+//using namespace Windows::Storage::Streams;
+
```
Comment (dargilco) [inline clients/cpp-console/include/WindowsMicMuter.h:18]: Remove above if not needed

File: clients/cpp-console/include/WindowsMicMuter.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#define SAFE_RELEASE(punk)  \
+              if ((punk) != NULL)  \
+                { (punk)->Release(); (punk) = NULL; }
+
```
Comment (dargilco) [inline clients/cpp-console/include/WindowsMicMuter.h:22]: punk? Use "ptr" or "x" instead.

File: clients/cpp-console/include/DeviceStatusIndicators.h [hunk 4]
```diff
@@ -5,6 +5,37 @@
...
+
+    template < class CharT, class Traits >
+    constexpr
+        std::basic_ostream< CharT, Traits >& foreground_yellow(std::basic_ostream< CharT, Traits >& os)
+    {
+        return os << "\033[33m";
+    }
+} // ansi
+
 enum class DeviceStatus
```
Comment (dargilco) [inline clients/cpp-console/include/DeviceStatusIndicators.h:39]: Does the above compile for Linux without errors?

File: clients/cpp-console/include/DialogManager.h [hunk 5]
```diff
@@ -56,6 +57,8 @@ class DialogManager
     void StopKws();
     // Start a listening session that read audio stream from a wav file.
     void ListenFromFile();
+    // Get mute state of the default microphone.
+    bool IsMuted() { return (_muter) ? _muter->IsMuted() : false; };
```
Comment (dargilco) [inline clients/cpp-console/include/DialogManager.h:61]: ? looks wrong
should be
{ return (_muter ? _muter->IsMuted() : false) };
I think you misplaced a closing parenthesis

File: clients/cpp-console/include/MicMuter.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    /// <remarks>
+    /// </remarks>
+    virtual ~IMicMuter() = default;
+
+    /// <summary>
+    /// Initialize will initialize the MicMuter with any specific OS dependent 
+    /// settings. If called without parameters it should assume some appropriate
+    /// defaults.
+    /// </summary>
+    /// <returns>A return code with S_OK as success and any other int as an error</returns>
```
Comment (dargilco) [inline clients/cpp-console/include/MicMuter.h:24]: S_OK is Windows specific. Say that a return value of 0 is success, and another other value is faulure

File: clients/cpp-console/include/MicMuter.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    /// <remarks>
+    /// </remarks>
+    virtual int Initialize() = 0;
+
+    /// <summary>
+    /// This method is used to actually mute and unmute the default microphone.
+    /// </summary>
+    /// <returns>A return code with S_OK as success and any other int as an error</returns>
```
Comment (dargilco) [inline clients/cpp-console/include/MicMuter.h:32]: Same as above

File: clients/cpp-console/src/windows/WindowsMicMuter.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+// Licensed under the MIT License.
+
+#include <atlbase.h>
+#include <string>
+#include <thread>
+#include "WindowsMicMuter.h"
+
+//#include <mutex>
+//#include <avrt.h>
+
```
[the remark this answers is not in the document]
Comment (qfweiwei) [reply clients/cpp-console/src/windows/WindowsMicMuter.cpp:11]: Updated.

File: clients/cpp-console/src/windows/WindowsMicMuter.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    SAFE_RELEASE(m_endpointVolume);
+}
+
+int WindowsMicMuter::Initialize()
+{
+    HRESULT hr = S_OK;
+    CComPtr<IMMDeviceEnumerator> pEnumerator;
+    CComPtr<IMMDevice> pDevice;
+
+    // begin Audio Device Setup
```
[the remark this answers is not in the document]
Comment (qfweiwei) [reply clients/cpp-console/src/windows/WindowsMicMuter.cpp:26]: Updated.

File: clients/cpp-console/src/windows/WindowsMicMuter.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+#include "WindowsMicMuter.h"
+
+//#include <mutex>
+//#include <avrt.h>
+
+using namespace MicMuter;
+
+WindowsMicMuter::~WindowsMicMuter()
+{
+    m_endpointVolume->SetMute(_originalMuteState, NULL);
```
[the remark this answers is not in the document]
Comment (qfweiwei) [reply clients/cpp-console/src/windows/WindowsMicMuter.cpp:16]: Updated.

File: clients/cpp-console/src/windows/WindowsMicMuter.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    hr = pDevice->Activate(__uuidof(IAudioEndpointVolume), CLSCTX_INPROC_SERVER, NULL,
+        reinterpret_cast<void**>(&m_endpointVolume));
+    if (hr != S_OK)
+    {
+        goto exit;
+    }
+    m_endpointVolume->GetMute(&_originalMuteState);
+    _muted = _originalMuteState == TRUE;
+
+exit:
```
[the remark this answers is not in the document]
Comment (qfweiwei) [reply clients/cpp-console/src/windows/WindowsMicMuter.cpp:56]: Updated.

File: clients/cpp-console/src/windows/WindowsMicMuter.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    else
+    {
+        hr = m_endpointVolume->SetMute(TRUE, NULL);
+    }
+
+    if (hr == S_OK)
+    {
+        _muted = !_muted;
+    }
+
```
[the remark this answers is not in the document]
Comment (qfweiwei) [reply clients/cpp-console/src/windows/WindowsMicMuter.cpp:78]: Updated.

File: clients/cpp-console/include/WindowsMicMuter.h (hunk 3 above)
Comment (qfweiwei) [reply clients/cpp-console/include/WindowsMicMuter.h:22]: Updated.

File: clients/cpp-console/include/DialogManager.h (hunk 5 above)
Comment (qfweiwei) [reply clients/cpp-console/include/DialogManager.h:61]: I like to use parenthesis this way like they are used after "if". It works here, but we can change.

File: clients/cpp-console/include/MicMuter.h (hunk 6 above)
Comment (qfweiwei) [reply clients/cpp-console/include/MicMuter.h:24]: Updated.

File: clients/cpp-console/include/MicMuter.h (hunk 7 above)
Comment (qfweiwei) [reply clients/cpp-console/include/MicMuter.h:32]: Updated.

**Pull request merged**